pkg/targetallocator/adapters/config_to_prom_config.go

```diff
@@ -57,3 +68,170 @@ func ConfigToPromConfig(cfg string) (map[interface{}]interface{}, error) {
 
 	return prometheus, nil
 }
+
+// UnescapeDollarSignsInPromConfig replaces "$$" with "$" in the "replacement" fields of
+// both "relabel_configs" and "metric_relabel_configs" in a Prometheus configuration file.
+func UnescapeDollarSignsInPromConfig(cfg string) (map[interface{}]interface{}, error) {
+	prometheus, err := ConfigToPromConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	prometheusConfigProperty, ok := prometheus["config"]
+	if !ok {
+		return nil, errorNoComponent("prometheusConfig")
+	}
+
+	prometheusConfig, ok := prometheusConfigProperty.(map[interface{}]interface{})
+	if !ok {
+		return nil, errorNotAMap("prometheusConfig")
+	}
+
+	scrapeConfigsProperty, ok := prometheusConfig["scrape_configs"]
+	if !ok {
+		return nil, errorNoComponent("scrape_configs")
+	}
+
+	scrapeConfigs, ok := scrapeConfigsProperty.([]interface{})
+	if !ok {
+		return nil, errorNotAList("scrape_configs")
+	}
+
+	for _, config := range scrapeConfigs {
+		scrapeConfig, ok := config.(map[interface{}]interface{})
+		if !ok {
+			return nil, errorNotAMap("scrape_config")
+		}
+
+		relabelConfigsProperty, ok := scrapeConfig["relabel_configs"]
+		if !ok {
+			continue
+		}
+
+		relabelConfigs, ok := relabelConfigsProperty.([]interface{})
+		if !ok {
+			return nil, errorNotAList("relabel_configs")
+		}
+
+		for _, rc := range relabelConfigs {
+			relabelConfig, rcErr := rc.(map[interface{}]interface{})
+			if !rcErr {
+				return nil, errorNotAMap("relabel_config")
+			}
+
+			replacementProperty, rcErr := relabelConfig["replacement"]
+			if !rcErr {
+				continue
+			}
+
+			replacement, rcErr := replacementProperty.(string)
+			if !rcErr {
+				return nil, errorNotAString("replacement")
+			}
+
+			relabelConfig["replacement"] = strings.ReplaceAll(replacement, "$$", "$")
```

Comment:
are we unable to just do the replace all on the string configuration we're given? Do we need to only run it on relabel_configs and metric_relabel_configs?

Reply:
@kristinapathak thinks we should maybe try and just regex this change as well.